Replace rlpvalue with oxenc; use oxenc for more optimized hex conversions; various code modernizations

jagerman opened the pull request:
- Replace rlpvalue with oxenc (which now has an rlp serializer)
- Use oxenc for hex conversions and hex literals
- Switch compilation to C++20
- Switch to std::string_views or std::spans in various places instead of const lvalue strings or specific vector/array types.
- Change Provider::sendTransaction and similar to not use `std::async` because they were just waiting for the async to return (which blocks the current thread anyway, and so might as well just happen in the current thread), and abstracted the get-and-wait code into a single function.
- removed some unused utils functions
- external -- brought in `system_or_submodule` from other oxen packages to have cmake check for system libs before building/using submodules for nlohman, secp256k1, oxen-encoding.
- Turn off `-Wshadow` because, under GCC, it produces warnings for things it shouldn't, like constructor arguments that initialize members, and lambdas that capture a copy of the same name.  Clang's `-Wshadow` is less stupid about this, but as this also causes a metric ton of warnings for dependent code (like oxend) just turn it off for now.
- Enable compile_commands.json generation by default
- Reduce header usage in the headers
- Various small cmake cleanups and tweaks

jagerman committed:
Reduce cpr, json header inclusions

Reduces cpr and json include weight in provider.hpp to just the types
needed in the header (cpr), and just the forward declaration (json).

De-templatized evm_increaseTime to avoid needing the full json headers.
Its template on a generic duration type isn't really useful (less
precise durations implicitly convert already, and more precise times
should be a call error, not silently truncated).

jagerman committed:
cmake: enable compile_commands.json export

This is cheap to generate, and extremely useful for editors that use
clangd for semantic completion. (And occassionally for humans who want
to see the exact flags being used for compilation, without having to
pull it out of the debug build log).

jagerman committed:
cmake: change src-only deps to PRIVATE linkage

ethyl-depending projects don't need them added to include paths/linkage.

jagerman committed:
Remove redundant (and case-insensitive) header

windows.h is already included at the top of this header.

jagerman committed:
Use oxenc; replace rlpvalue; c++20 modernizations and various cleanups

- Replace rlpvalue with oxenc (which now has an rlp serializer)
- Use oxenc for hex conversions and hex literals
- Switch compilation to C++20
- Switch to std::string_views or std::spans in various places instead of
  const lvalue strings or specific vector/array types.
- Change Provider::sendTransaction and similar to not use `std::async`
  because they were just waiting for the async to return (which blocks
  the current thread anyway, and so might as well just happen in the
  current thread), and abstracted the get-and-wait code into a single
  function.
- removed some unused utils functions
- external -- brought in `system_or_submodule` from other oxen packages
  to have cmake check for system libs before building/using submodules
  for nlohman, secp256k1, oxen-encoding.
- Turn off `-Wshadow` because, under GCC, it produces warnings for
  things it shouldn't, like constructor arguments that initialize
  members, and lambdas that capture a copy of the same name.  Clang's
  `-Wshadow` is less stupid about this, but as this also causes a metric
  ton of warnings for dependent code (like oxend) just turn it off for
  now.

jagerman commented on CMakeLists.txt:
```diff
@@ -18,12 +21,6 @@ endif()
-    set(${PROJECT_NAME}_WARNINGS_AS_ERRORS ON CACHE BOOL "" FORCE)
-else()
-    set(${PROJECT_NAME}_ENABLE_UNIT_TESTING FALSE CACHE BOOL "" FORCE)
-endif()
```
(These now get defaulted to ON/OFF later on based on being top-level, rather than unconditionally forced on/off here)

jagerman committed:
Bump oxenc for cmake C++20 compilation propagation

jagerman committed:
Re-add padToNBytes

jagerman committed:
Don't add oxenc/nlohmann if a parent project is providing it

darcys22 commented on include/ethyl/utils.hpp:
```diff
@@ -38,23 +43,72 @@ namespace utils
+
+    std::string padToNBytes(std::string_view hex_input, size_t bytes, PaddingDirection direction = PaddingDirection::LEFT);
+    inline std::string padTo8Bytes(std::string_view hex_input, PaddingDirection direction = PaddingDirection::LEFT) {
+        return padToNBytes(hex_input, 32, direction);
```
pads to 32 bytes when it should be padding to 8 bytes. Looks like the other function has it swapped also

jagerman replied:
Fixed.

darcys22 commented:
There is something which is causing this 
```
due to unexpected exception with message:                                                                                                                         
  Error sending transaction: Error: s cannot have leading zeroes, received:                                                                                       
  0081805ed9683781a7320042fed7f86d529dee82107df305eb8a21d727a785e9 
```
  
Still tracking it down but I think this is happening for every signed transaction. It only happens when the first digits of the signature of the transaction are zero so its randomly failing to send transactions

darcys22 commented on src/transaction.cpp:
```diff
@@ -1,50 +1,39 @@
+    if (!sig.isEmpty()) {
+        arr.push_back(sig.signatureYParity);
+        arr.push_back(sig.signatureR);
+        arr.push_back(sig.signatureS);
```
I think this has now lost
```
            appendDataToRLP(arr, sig.signatureR, utils::removeLeadingZeros);
            appendDataToRLP(arr, sig.signatureS, utils::removeLeadingZeros);
```
and cause the leading zeros arnt being trimmed its causing transactions to fail occasionally

darcys22 replied:
```
        arr.push_back(utils::removeLeadingZeros(sig.signatureR));
        arr.push_back(utils::removeLeadingZeros(sig.signatureS));
```

jagerman replied:
Pushed a fix for this (which moves this remove-leading-zeros logic back into oxenc -- it seems more appropriate there as it's an encoding implementation detail for big integer values).

jagerman committed:
Fix reversed 8/32 padding

jagerman committed:
Move 0x integer prefix into decimalToHex

Adds an extra `bool` argument that, when true, prepends the 0x to the
returned value.

jagerman committed:
Use oxenc to fix big integer RLP encoding of signature values

Remove the unused intToBytes and removeLeadingZeros functions as these
were only used for RLP encoding implementation details (handled by
oxenc's RLP encoding code).

darcys22 commented:
So in the serializing transaction code its doing this:

```
Incorrect
"0x02
e70101808082520894
23abf16ede550000f45279dcbea84f56c6cebc80
880de0b6b3a764000080c0"


correct
"0x02
e70101808082520894
a6c077fd9283421c657ecea8a9c1422cc6cebc80 <- This is the address of the transaction
880de0b6b3a764000080c0"
```

The address is what appears to be serialized incorrectly in the transaction. However the last 4 bytes/8 chars are correct.

The full address is 20 bytes/40 characters. The first 16 bytes/ 32 characters may be getting overwritten somehow?

Edit: Looks like the 16 bytes that are different are changing every run too

darcys22 committed:
fromHexString returns a vector and was being converted to a span
